client/helpers.go

```diff
@@ -382,3 +390,39 @@ func TagsToMap(tagSlice interface{}) map[string]string {
 	TagsIntoMap(tagSlice, ret)
 	return ret
 }
+
+func ListAndDetailResolver(ctx context.Context, meta schema.ClientMeta, res chan<- interface{}, list ListResolver, details DetailResolver) error {
+	var diags diag.Diagnostics
+	var wg sync.WaitGroup
+	sem := semaphore.NewWeighted(int64(MAX_GOROUTINES))
+	errorChan := make(chan error)
+	response, err := list(ctx, meta)
+	if err != nil {
+		return diag.WrapError(err)
+	}
+	// All items will be attempted to be fetched, but could return an error
+	wg.Add(len(response))
+	for _, item := range response {
+		if err := sem.Acquire(ctx, 1); err != nil {
+			return diag.WrapError(err)
+		}
+		func(summary interface{}) {
+			defer wg.Done()
+			defer sem.Release(1)
+			details(ctx, meta, res, errorChan, summary)
+		}(item)
+	}
+	// Ensure all items have been attempted to be fetched
+	wg.Wait()
+
+	// empty the error channel and convert to Diags
+	for len(errorChan) > 0 {
```

Comment:
why not do a range? you can `close(errorChan)` after `wg.Wait` returns.

Reply:
you'd need to use a buffered channel and preallocate (limited) "slots" for potential errors.

Also since errorChan is unbuffered you even need to start reading in a separate goroutine before firing off the first worker... then `range` will stop reading once the channel is closed. See https://github.com/cloudquery/cq-provider-sdk/pull/208/files for an example.

Reply:
The changes look good 👍